site/_posts/2020-05-20-building-protected-api-with-rails-and-jwt.md

```diff
@@ -0,0 +1,254 @@
+
+```json
+{"messages":["Hello"]}
+```
+
+But let's add a test so future changes don't cause any surprises. Create the controller test at `test/controllers/messages_controller_test.rb`. Here are the contents of that class:
+
+```ruby
+require 'test_helper'
+
+class MessagesTest < ActionDispatch::IntegrationTest
+  test "can get messages" do
+    get "/messages"
+    assert_response :success
+  end
+  test "can get messages content" do
+    get "/messages"
+    res = JSON.parse(@response.body)
+    assert_equal '{"messages"=>["Hello"]}', res.to_s
+  end
+end
+```
+
+Now we can run our test and make sure that we are getting what we expect:
+
+```shell
+$ rails test test/integration/messages_test.rb
+Running via Spring preloader in process 15492
+Run options: --seed 1452
+
+# Running:
+
+..
+
+Finished in 0.119373s, 16.7542 runs/s, 16.7542 assertions/s.
+2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
+```
+
+Excellent! We have a working API which returns well formed JSON! Rails even takes care of setting the `Content-Type` header to `application/json; charset=utf-8`. Now let's secure the API.
+
+## Secure the API
+
+As a reminder, we're going to use a JWT to secure this API. While you can secure Rails APIs using [a variety of methods](https://edgeguides.rubyonrails.org/action_controller_overview.html#http-authentications), using a token has advantages. You can integrate with a number of identity providers that offer OAuth or SAML support, which allows you to leverage an existing robust identity management system to control API access. You can also embed additional metadata into a JWT, including attributes like roles.
+
+The first step to changing this API is to write tests. Let's modify the test to provide a JWT and expect `:forbidden` HTTP statuses when the token doesn't meet our expectations.
+
+```ruby
+class MessagesTest < ActionDispatch::IntegrationTest
+  test "can' get messages with no auth" do
+    get "/messages"
+    assert_response :forbidden
+  end
+  test "can get messages with header" do
+    get "/messages", headers: { "HTTP_AUTHORIZATION" => "Bearer " + build_jwt }
+    assert_response :success
+  end
+  test "expired jwt fails" do
+    get "/messages", headers: { "HTTP_AUTHORIZATION" => "Bearer " + build_jwt(-1) }
+    assert_response :forbidden
+  end
+  test "can get messages content" do
+    get "/messages", headers: { "HTTP_AUTHORIZATION" => "Bearer " + build_jwt }
+    res = JSON.parse(@response.body)
+    assert_equal '{"messages"=>["Hello"]}', res.to_s
+  end
+
+  def build_jwt(valid_for_minutes = 5)
+    exp = Time.now.to_i + (valid_for_minutes*60)
+    payload = { "iss": "fusionauth.io",
+                "exp": exp,
+                "aud": "238d4793-70de-4183-9707-48ed8ecd19d9",
+                "sub": "19016b73-3ffa-4b26-80d8-aa9287738677",
+                "name": "Dan Moore",
+                "roles": ["USER"]
+    }
+
+    JWT.encode payload, Rails.configuration.x.oauth.jwt_secret, 'HS256'
```

Comment:
Is this a library that you are using? We might want to explain where this comes from and how to install it if that is the case.

Reply:
Yup, it comes with rails, though. I'll note that it's a library that is included.

Reply:
Looks like it wasn't included, so I added install steps.